# IBSplot

## Count variations using different windows sizes:

	Use the script: var_byWind.py
	Or adjust the bash script to your folder paths and arguments: run_var_byWind.sh

### Run the code in the command line to generate a 500,000 bp windows file from the 50,000 bp original file:
```sh
	python var_byWind.py \
		-db kmeribs-Wheat_Jagger-Flame.tsv.gz.gz \
		-cl chr_sizes_jagger.genome.txt \
		-w 500000 \
		-o kmeribs-Wheat_Jagger-Flame_500000.tsv.gz
```
### Arguments description:
```
	-db [ output ".tsv.gz" file generated by the IBSpy analysis ]
	-cl [ tab separated no header file with column one; chromosome names and column two; chromosome lengths]
	-w [ size of the windows to count variations ]
	-o [ file output ".tsv" with three columns "seqname", "start", "variations"; this can be used by the HapGMMbuild.py script exactly as described above ]
```

```
	chromosome lengths example of the reference jagger: chr_sizes_jagger.genome.txt

chr1A	596211899
chr1B	705338699
chr1D	493450010
chr2A	804285258
chr2B	810734643
chr2D	673981989
chr3A	754496630
chr3B	855759449
chr3D	614042580
chr4A	743847818
chr4B	673340788
chr4D	515668560
chr5A	713320806
chr5B	703598484
chr5D	570159854
chr6A	625793224
chr6B	721110502
chr6D	459355444
chr7A	745201001
chr7B	749284433
chr7D	643344672
chrUn	379824255
```


## Build haplotypes using the 500,000 bp windows using the file created above:

	Use the python script: HapGMMbuild.py
	Or adjust the bash script to your folder paths and arguments: run_HapGMMbuild.sh

### Run the code in the command line as the follow example for 500,000 bp windows and 1800 variation count filter using Jagger and Flame samples:
NOTE: you will need to adjust the windows and filters according to the input data of the reference and query samples.

```sh
	python HapGMMbuild.py \
		-db kmeribs-Wheat_Jagger-Flame_500000.tsv.gz \
		-rf jagger \
		-qr Flame \
		-w 500000 \
		-vf 1800 \
		-nc 3 \
		-chr chr1A \
		-st 3 \
		-gm gmm_Jagger_vs_Flame_500000_wd_1800_flt.tsv.gz \
		-ho chr1A_Jagger_vs_Flame_500000_wd_1800_flt.tsv.gz

```
### Arguments description:
```
	-db [ output ".tsv.gz" file generated by the var_byWind.py above script ]
	-rf [ genome reference used in the IBSpy (Jagger in this example) analysis ]
	-qr [ name of the query or test sample used in the IBSpy analysis (Flame in this example) ]
	-w [ size of the windows to count variations ]
	-vf [ filter variations above this threshold to compute GMM model ]
	-nc [ number of components used in the GMM model, we are mainly using 3, but this can be changed according to the samples used ]
	-chr [ chromosome name to plot; must match the name in the "-cl" argument file, column one ]
	-st [consecutive "outliers" in windows to stitch]
	-gm [ output ".tsv.gz" file categorized in the column "gauss_mx_IBS" as a IBS == 1 and non-IBS == 0 ]
	-ho [ haplotypes stitched using the GMM model output file grouped by IBS (1) and non-IBS (0) ]
```


## Generate histograms to visualize variation count distributions and categories generated by the GMM model:

	Use the script: plot_HapGMM.py
	Or adjust the bash script to your folder paths and arguments: run_histo_HapGMM.sh

```sh
	python histo_GMM.py \
		-db kmeribs-Wheat_Jagger-Flame_500000.tsv.gz \
		-gm gmm_Jagger_vs_Flame_500000_wd_1800_flt.tsv.gz \
		-rf jagger \
		-qr Flame \
		-w 500000 \
		-vf 1800 \
		-do hist_Jagger_vs_Flame_500000_wd_1800_flt.pdf \
		-go hist_Jagger_vs_Flame_500000_wd_whole_genome.pdf
```
### Arguments description:
```
Same as "HapGMMbuild.py", but with the additional following arguments: 
	-do [ Output histogram of the variations from the whole genome in ".pdf" format ]
	-ho [ Output histogram of the variations from the whole genome by IBS and non-IBS in ".pdf" format ]
```


## Plot final haploblocks:

	Use the script: plot_HapGMM.py
	Or adjust the bash script to your folder paths and arguments: run_plot_HapGMM.sh
	
### Run the code in the command line as follow to generate haplotypes using 500,000 bp windows and filter for 1800 count using jagger and flame samples:
```sh
	python plot_HapGMM.py \
		-ho chr1A_Jagger_vs_Flame_500000_wd_1800_flt.tsv \
		-cl chr_sizes_jagger.genome.txt \
		-rf jagger \
		-qr Flame \
		-chr chr1A \
		-w 500000 \
		-vf 1800 \
		-o hap_blocks_chr1A_Jagger_vs_Flame_500000_wd_1800_flt.jpg
```

### Arguments description:
```
	-ho [ file generated by the HapGMMbuild.py described above with the -gm command, which will have the "gauss_mx_IBS" column with 0/1 ]
	-cl [ tab separated no header file with column one; chromosome names and column two; chromosome lengths (described above)]
	-rf [ genome reference used in the IBSpy analysis ]
	-qr [ name of the query or test sample used in the IBSpy analysis ]
	-chr [ chromosome name to plot; must match the name in the "-cl" argument file, column one]
	-w [ size of the windows to count variations; must match the windows used to compute the GMM model ]
	-vf [ filter variations above this threshold to compute GMM model; must match the windows used to compute the GMM model ]
	-o [ output scatter plot and bars with haplotypes in ".jpg" format ]

```
end...