# IBSplot

## Build haplotypes directly from the IBSpy output ".tsv.gz" file using the 100,000 bp windows:

	Use the python script: HapGMMbuild.py
	Or adjust the bash script to your folder paths and arguments: run_HapGMMbuild.sh

### Run the code in the command line as the follow example for 100,000 bp windows and 400 count using symattis and flame samples:
NOTE: you will need to adjust the windows and filters according to the input data of the reference and query samples.

```sh
	python HapGMMbuild.py \
		-db flame_symattis_100000.tsv.gz \
		-rf symattis \
		-qr flame-10x_k31_agis \
		-w 100000 \
		-vf 400 \
		-nc 3 \
		-gmf gmm_flame_symattis_100000_wd_400_flt_agis.tsv \
		-hr all_var_100000.pdf \
		-hg gmm_histo_plot_categorized_100000.pdf
```
### Arguments description:
```
	-db [ output ".tsv.gz" file generated by the IBSpy analysis ]
	-rf [ genome reference used in the IBSpy analysis ]
	-qr [ name of the query or test sample used in the IBSpy analysis ]
	-w [ size of the windows to count variations ]
	-vf [ filter variations above this threshold to compute GMM model ]
	-nc [ number of components used in the GMM model, we are mainly using 3, but this can be changed according to the samples used ]
	-gmf [ output .tsv file categorized in the column "gauss_mx_IBS" as a IBS == 1 and non-IBS == 0]
	-hr [ histogram of all variants from the whole genome ]
	-hg [ histogram if the GMM categorized data split by group ]
```

## Count variations using different windows size:

	Use the script: var_byWind.py
	Or adjust the bash script to your folder paths and arguments: run_var_byWind.sh

### Run the code in the command line to generate a 500,000 bp windows file from the 100,000 bp original file:
```sh
	python var_byWind.py \
		-db flame_symattis_100000.tsv.gz \
		-cl chr_sizes_symattis.genome.txt \
		-w 500000 \
		-o flame_symattis_500000.tsv
```

### Arguments description:
```
	-db [ output ".tsv.gz" file generated by the IBSpy analysis ]
	-cl [ tab separated no header file with column one; chromosome names and column two; chromosome lengths]
	-w [ size of the windows to count variations ]
	-o [ file output ".tsv" with three columns "seqname", "start", "variations"; this can be used by the HapGMMbuild.py script exactly as described above ]
```
```
	chromosome lengths example of the reference symattis: chr_sizes_symattis.genome.txt

	chr1A	600654286
	chr1B	685083685
	chr1D	486317123
	chr2A	794150360
	chr2B	799857935
	chr2D	655329108
	chr3A	749763888
	chr3B	838116175
	chr3D	610468321
	chr4A	736551279
	chr4B	666328382
	chr4D	504826275
	chr5A	702606209
	chr5B	467876140
	chr5D	566465558
	chr6A	614421429
	chr6B	698878671
	chr6D	480431564
	chr7A	735408736
	chr7B	969998116
	chr7D	635024734
	chrUn	956807450
```


## Build final haplotypes and plot:

	Use the script: plot_HapGMM.py
	Or adjust the bash script to your folder paths and arguments: run_plot_HapGMM.sh
	
### Run the code in the command line as follow to generate haplotypes using 500,000 bp windows and filter for 400 count using symattis and flame samples:
```sh
	python plot_HapGMM.py \
		-gm gmm_symattis_vs_flame_500000_wd_400_flt.tsv \
		-db flame_symattis_500000.tsv \
		-cl chr_sizes_symattis.genome.txt \
		-rf symattis \
		-qr flame \
		-chr chr6A \
		-w 500000 \
		-vf 400 \
		-st 3 \
		-o hap_blocks_chr6A_symattis_vs_flame_500000_wd_400_flt.jpg
```

### Arguments description:
```
	-gm [ file generated by the HapGMMbuild.py described above with the -gmf command, which will have the "gauss_mx_IBS" column with 0/1 ]
	-db [ output ".tsv.gz" file generated by the IBSpy analysis, this is used to get the same coordinates when plotting ]
	-cl [ tab separated no header file with column one; chromosome names and column two; chromosome lengths (described above)]
	-rf [ genome reference used in the IBSpy analysis ]
	-qr [ name of the query or test sample used in the IBSpy analysis ]
	-chr [ chromosome name to plot; must match the name in the "-cl" argument file, column one]
	-w [ size of the windows to count variations; must match the windows used to compute the GMM model ]
	-vf [ filter variations above this threshold to compute GMM model; must match the windows used to compute the GMM model ]
	-st [ stitching haplotypes: number of non-IBS "outliers" that must a appear consecutively in a windows to be called non-IBS ]
	-o [ output scatter plot and bars with haplotypes in ".jpg" format ]

```

