{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "automated-wheat",
   "metadata": {},
   "source": [
    "# Scoring of variations\n",
    "\n",
    "There are three different cases:\n",
    "\n",
    "1. Simple SNPs\n",
    "2. Indels on the query\n",
    "3. SNPs at a distance less than  K\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "surprised-ceramic",
   "metadata": {},
   "outputs": [],
   "source": [
    "from IBSpy import KmerGWASDBBuilder, KmerGWASDB, FastaChunkReader \n",
    "from pyfaidx import Fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "relative-hartford",
   "metadata": {},
   "outputs": [],
   "source": [
    "kmerdb  = KmerGWASDB(31)\n",
    "kmerdb.load_from_fasta(\"../tests/data/test4B.jagger.fa\")\n",
    "builder = KmerGWASDBBuilder(31)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "quality-prescription",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_iter = FastaChunkReader(\"../tests/data/test4B.stanley.rev.fa\", kmer_size = 31)\n",
    "sequence_sta = list(fasta_iter )[0]\n",
    "fasta_iter = FastaChunkReader(\"../tests/data/test4B.jagger.fa\", kmer_size = 31)\n",
    "sequence_jag = list(fasta_iter )[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "violent-mission",
   "metadata": {},
   "source": [
    "## Simple case (SNP)\n",
    "\n",
    "A single SNP is recorded as a single variation, a kmer_distance of 1 and the observed kmers is $total_kmers - k$. In this example, 33 kmers, with a SNP in the centre only display 2 observed kmers. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "recorded-caribbean",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cctcgcctccgcgacctggagcgattgctggTcgtcggctatgtgcccgcgcagcgccgtgcc\n",
      "cctcgcctccgcgacctggagcgattgctggGcgtcggctatgtgcccgcgcagcgccgtgcc\n",
      "CCTCGCCTCCGCGACCTGGAGCGATTGCTGG\n",
      " ctcgcctccgcgacctggagcgattgctggG\n",
      "  tcgcctccgcgacctggagcgattgctggGc\n",
      "   cgcctccgcgacctggagcgattgctggGcg\n",
      "    gcctccgcgacctggagcgattgctggGcgt\n",
      "     cctccgcgacctggagcgattgctggGcgtc\n",
      "      ctccgcgacctggagcgattgctggGcgtcg\n",
      "       tccgcgacctggagcgattgctggGcgtcgg\n",
      "        ccgcgacctggagcgattgctggGcgtcggc\n",
      "         cgcgacctggagcgattgctggGcgtcggct\n",
      "          gcgacctggagcgattgctggGcgtcggcta\n",
      "           cgacctggagcgattgctggGcgtcggctat\n",
      "            gacctggagcgattgctggGcgtcggctatg\n",
      "             acctggagcgattgctggGcgtcggctatgt\n",
      "              cctggagcgattgctggGcgtcggctatgtg\n",
      "               ctggagcgattgctggGcgtcggctatgtgc\n",
      "                tggagcgattgctggGcgtcggctatgtgcc\n",
      "                 ggagcgattgctggGcgtcggctatgtgccc\n",
      "                  gagcgattgctggGcgtcggctatgtgcccg\n",
      "                   agcgattgctggGcgtcggctatgtgcccgc\n",
      "                    gcgattgctggGcgtcggctatgtgcccgcg\n",
      "                     cgattgctggGcgtcggctatgtgcccgcgc\n",
      "                      gattgctggGcgtcggctatgtgcccgcgca\n",
      "                       attgctggGcgtcggctatgtgcccgcgcag\n",
      "                        ttgctggGcgtcggctatgtgcccgcgcagc\n",
      "                         tgctggGcgtcggctatgtgcccgcgcagcg\n",
      "                          gctggGcgtcggctatgtgcccgcgcagcgc\n",
      "                           ctggGcgtcggctatgtgcccgcgcagcgcc\n",
      "                            tggGcgtcggctatgtgcccgcgcagcgccg\n",
      "                             ggGcgtcggctatgtgcccgcgcagcgccgt\n",
      "                              gGcgtcggctatgtgcccgcgcagcgccgtg\n",
      "                               Gcgtcggctatgtgcccgcgcagcgccgtgc\n",
      "                                CGTCGGCTATGTGCCCGCGCAGCGCCGTGCC\n",
      "seqname\tstart\tend\ttotal_kmers\tobserved_kmers\tvariations\tkmer_distance\n",
      "jag1\t0\t32\t33\t2\t1\t1\n"
     ]
    }
   ],
   "source": [
    "jag1  = sequence_jag[\"seq\"][4516:4579].lower() \n",
    "sta1  = sequence_sta[\"seq\"][4518:4581].lower()\n",
    "position = 31\n",
    "jag1 = jag1[:position] + jag1[position].upper() + jag1[position+1:]\n",
    "sta1 = sta1[:position] + sta1[position].upper() + sta1[position+1:]\n",
    "query = list(builder.sequence_to_kmers(sta1, convert=True))\n",
    "print(jag1)\n",
    "print(sta1)\n",
    "i = 0\n",
    "count = 0 \n",
    "for q in query:\n",
    "    q_str = builder.kmer_to_string(q)\n",
    "    q_str = \" \"*i + q_str\n",
    "    if(q not in kmerdb):\n",
    "        q_str = q_str.lower()\n",
    "        q_str = q_str[:position] + q_str[position].upper() + q_str[position+1:]\n",
    "    print(q_str)\n",
    "    i += 1\n",
    "\n",
    "stats = kmerdb.kmers_stats_from_sequence(query)\n",
    "stats.seqname = \"jag1\"\n",
    "stats.start   = 0\n",
    "stats.end     = len(query) - 1\n",
    "print(stats.header())\n",
    "print(stats.csv())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "everyday-carrier",
   "metadata": {},
   "source": [
    "## Indels\n",
    "\n",
    "Indels produce a single variation, a kmer distance of the size of the deletion and the total number of kmers is reduced by k plus the deleiton size\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "grand-partner",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTGCATTAATTCATTTAGATCTTTTGACACA--GGCAGCAGTTCCCAAGATTCTTTTGG\n",
      "TTGCATTAATTCATTTAGATCTTTTGACACACAGGCAGCAGTTCCCAAGATTCTTTTGG\n",
      "TTGCATTAATTCATTTAGATCTTTTGACACA\n",
      " tgcattaattcatttagatcttttgacacac\n",
      "  gcattaattcatttagatcttttgacacaca\n",
      "   cattaattcatttagatcttttgacacacag\n",
      "    attaattcatttagatcttttgacacacagg\n",
      "     ttaattcatttagatcttttgacacacaggc\n",
      "      taattcatttagatcttttgacacacaggca\n",
      "       aattcatttagatcttttgacacacaggcag\n",
      "        attcatttagatcttttgacacacaggcagc\n",
      "         ttcatttagatcttttgacacacaggcagca\n",
      "          tcatttagatcttttgacacacaggcagcag\n",
      "           catttagatcttttgacacacaggcagcagt\n",
      "            atttagatcttttgacacacaggcagcagtt\n",
      "             tttagatcttttgacacacaggcagcagttc\n",
      "              ttagatcttttgacacacaggcagcagttcc\n",
      "               tagatcttttgacacacaggcagcagttccc\n",
      "                agatcttttgacacacaggcagcagttccca\n",
      "                 gatcttttgacacacaggcagcagttcccaa\n",
      "                  atcttttgacacacaggcagcagttcccaag\n",
      "                   tcttttgacacacaggcagcagttcccaaga\n",
      "                    cttttgacacacaggcagcagttcccaagat\n",
      "                     ttttgacacacaggcagcagttcccaagatt\n",
      "                      tttgacacacaggcagcagttcccaagattc\n",
      "                       ttgacacacaggcagcagttcccaagattct\n",
      "                        tgacacacaggcagcagttcccaagattctt\n",
      "                         gacacacaggcagcagttcccaagattcttt\n",
      "                          acacacaggcagcagttcccaagattctttt\n",
      "                           cacacaggcagcagttcccaagattcttttg\n",
      "                            ACACAGGCAGCAGTTCCCAAGATTCTTTTGG\n",
      "seqname\tstart\tend\ttotal_kmers\tobserved_kmers\tvariations\tkmer_distance\n",
      "jag1\t0\t28\t29\t2\t1\t2\n"
     ]
    }
   ],
   "source": [
    "jag1  = sequence_jag[\"seq\"][1948:2005]\n",
    "sta1  = sequence_sta[\"seq\"][1948:2007]\n",
    "position = 31\n",
    "jag1 = jag1[:position] + \"--\" + jag1[position:]\n",
    "query = list(builder.sequence_to_kmers(sta1, convert=True))\n",
    "print(jag1)\n",
    "print(sta1)\n",
    "i = 0\n",
    "count = 0 \n",
    "for q in query:\n",
    "    q_str = builder.kmer_to_string(q)\n",
    "    q_str = \" \"*i + q_str\n",
    "    if(q not in kmerdb):\n",
    "        q_str = q_str.lower()\n",
    "        #q_str = q_str[:position] + q_str[position].upper() + q_str[position+1:]\n",
    "    print(q_str)\n",
    "    i += 1\n",
    "\n",
    "stats = kmerdb.kmers_stats_from_sequence(query)\n",
    "stats.seqname = \"jag1\"\n",
    "stats.start   = 0\n",
    "stats.end     = len(query) - 1\n",
    "print(stats.header())\n",
    "print(stats.csv())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "operational-import",
   "metadata": {},
   "source": [
    "## SNPs closer at a distance less than K \n",
    "\n",
    "When SNPs are close by, they are registered as a single variation. However, the kmer_distance of the event is longer, as all the kmers between the SNPs are missing. In this case, the kmer_distance is calculated by the distance between the first and the last SNP plus one. If there were more SNPs in between, the distance would still be the same. This hides regions with a different level of conservation, but we can't resolve it just with the kmer database. We need the connections between kmers to resolve the path.   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "mighty-pipeline",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "acaaaaggcgagtaaaagagttctctgttcaCcggctacgagcagcgtagacaaaggaagTggcggcgataacagtcgtggtagggcacagc\n",
      "acaaaaggcgagtaaaagagttctctgttcaAcggctacgagcagcgtagacaaaggaagCggcggcgataacagtcgtggtagggcacagc\n",
      "ACAAAAGGCGAGTAAAAGAGTTCTCTGTTCA\n",
      " caaaaggcgagtaaaagagttctctgttcaA\n",
      "  aaaaggcgagtaaaagagttctctgttcaAc\n",
      "   aaaggcgagtaaaagagttctctgttcaAcg\n",
      "    aaggcgagtaaaagagttctctgttcaAcgg\n",
      "     aggcgagtaaaagagttctctgttcaAcggc\n",
      "      ggcgagtaaaagagttctctgttcaAcggct\n",
      "       gcgagtaaaagagttctctgttcaAcggcta\n",
      "        cgagtaaaagagttctctgttcaAcggctac\n",
      "         gagtaaaagagttctctgttcaAcggctacg\n",
      "          agtaaaagagttctctgttcaAcggctacga\n",
      "           gtaaaagagttctctgttcaAcggctacgag\n",
      "            taaaagagttctctgttcaAcggctacgagc\n",
      "             aaaagagttctctgttcaAcggctacgagca\n",
      "              aaagagttctctgttcaAcggctacgagcag\n",
      "               aagagttctctgttcaAcggctacgagcagc\n",
      "                agagttctctgttcaAcggctacgagcagcg\n",
      "                 gagttctctgttcaAcggctacgagcagcgt\n",
      "                  agttctctgttcaAcggctacgagcagcgta\n",
      "                   gttctctgttcaAcggctacgagcagcgtag\n",
      "                    ttctctgttcaAcggctacgagcagcgtaga\n",
      "                     tctctgttcaAcggctacgagcagcgtagac\n",
      "                      ctctgttcaAcggctacgagcagcgtagaca\n",
      "                       tctgttcaAcggctacgagcagcgtagacaa\n",
      "                        ctgttcaAcggctacgagcagcgtagacaaa\n",
      "                         tgttcaAcggctacgagcagcgtagacaaag\n",
      "                          gttcaAcggctacgagcagcgtagacaaagg\n",
      "                           ttcaAcggctacgagcagcgtagacaaagga\n",
      "                            tcaAcggctacgagcagcgtagacaaaggaa\n",
      "                             caAcggctacgagcagcgtagacaaaggaag\n",
      "                              aAcggctacgagcagcgtagacaaaggaagC\n",
      "                               AcggctacgagcagcgtagacaaaggaagCg\n",
      "                                cggctacgagcagcgtagacaaaggaagCgg\n",
      "                                 ggctacgagcagcgtagacaaaggaagCggc\n",
      "                                  gctacgagcagcgtagacaaaggaagCggcg\n",
      "                                   ctacgagcagcgtagacaaaggaagCggcgg\n",
      "                                    tacgagcagcgtagacaaaggaagCggcggc\n",
      "                                     acgagcagcgtagacaaaggaagCggcggcg\n",
      "                                      cgagcagcgtagacaaaggaagCggcggcga\n",
      "                                       gagcagcgtagacaaaggaagCggcggcgat\n",
      "                                        agcagcgtagacaaaggaagCggcggcgata\n",
      "                                         gcagcgtagacaaaggaagCggcggcgataa\n",
      "                                          cagcgtagacaaaggaagCggcggcgataac\n",
      "                                           agcgtagacaaaggaagCggcggcgataaca\n",
      "                                            gcgtagacaaaggaagCggcggcgataacag\n",
      "                                             cgtagacaaaggaagCggcggcgataacagt\n",
      "                                              gtagacaaaggaagCggcggcgataacagtc\n",
      "                                               tagacaaaggaagCggcggcgataacagtcg\n",
      "                                                agacaaaggaagCggcggcgataacagtcgt\n",
      "                                                 gacaaaggaagCggcggcgataacagtcgtg\n",
      "                                                  acaaaggaagCggcggcgataacagtcgtgg\n",
      "                                                   caaaggaagCggcggcgataacagtcgtggt\n",
      "                                                    aaaggaagCggcggcgataacagtcgtggta\n",
      "                                                     aaggaagCggcggcgataacagtcgtggtag\n",
      "                                                      aggaagCggcggcgataacagtcgtggtagg\n",
      "                                                       ggaagCggcggcgataacagtcgtggtaggg\n",
      "                                                        gaagCggcggcgataacagtcgtggtagggc\n",
      "                                                         aagCggcggcgataacagtcgtggtagggca\n",
      "                                                          agCggcggcgataacagtcgtggtagggcac\n",
      "                                                           gCggcggcgataacagtcgtggtagggcaca\n",
      "                                                            Cggcggcgataacagtcgtggtagggcacag\n",
      "                                                             GGCGGCGATAACAGTCGTGGTAGGGCACAGC\n",
      "seqname\tstart\tend\ttotal_kmers\tobserved_kmers\tvariations\tkmer_distance\n",
      "jag1\t0\t61\t62\t2\t1\t30\n"
     ]
    }
   ],
   "source": [
    "jag1  = sequence_jag[\"seq\"][3348:3440].lower()\n",
    "sta1  = sequence_sta[\"seq\"][3350:3442].lower()\n",
    "position = 31\n",
    "jag1 = jag1[:position] + jag1[position].upper() + jag1[position+1:]\n",
    "sta1 = sta1[:position] + sta1[position].upper() + sta1[position+1:]\n",
    "position2 = 31 + 29\n",
    "jag1 = jag1[:position2] + jag1[position2].upper() + jag1[position2+1:]\n",
    "sta1 = sta1[:position2] + sta1[position2].upper() + sta1[position2+1:]\n",
    "query = list(builder.sequence_to_kmers(sta1, convert=True))\n",
    "print(jag1)\n",
    "print(sta1)\n",
    "i = 0\n",
    "count = 0 \n",
    "for q in query:\n",
    "    q_str = builder.kmer_to_string(q)\n",
    "    q_str = \" \"*i + q_str\n",
    "    if(q not in kmerdb):\n",
    "        q_str = q_str.lower()\n",
    "        q_str = q_str[:position] + q_str[position].upper() + q_str[position+1:]\n",
    "        if i > 29:\n",
    "            q_str = q_str[:position2] + q_str[position2].upper() + q_str[position2+1:]\n",
    "    print(q_str)\n",
    "    i += 1\n",
    "\n",
    "stats = kmerdb.kmers_stats_from_sequence(query)\n",
    "stats.seqname = \"jag1\"\n",
    "stats.start   = 0\n",
    "stats.end     = len(query) - 1\n",
    "print(stats.header())\n",
    "print(stats.csv())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
